File: core/ml/angles.py
"""
Pose landmarks → joint angles (degrees).

Each angle is defined by three joints: A - VERTEX - C.
The returned angle is always at the VERTEX, between 0° and 180°.
"""
from __future__ import annotations

from typing import Dict, List

import numpy as np

from .extractor import LANDMARK_NAMES, PoseFrame

# (joint_a, vertex, joint_c) — angle measured at vertex
ANGLE_DEFS: Dict[str, tuple[str, str, str]] = {
    'left_knee':      ('left_hip',       'left_knee',      'left_ankle'),
    'right_knee':     ('right_hip',      'right_knee',     'right_ankle'),
    'left_hip':       ('left_shoulder',  'left_hip',       'left_knee'),
    'right_hip':      ('right_shoulder', 'right_hip',      'right_knee'),
    'left_elbow':     ('left_shoulder',  'left_elbow',     'left_wrist'),
    'right_elbow':    ('right_shoulder', 'right_elbow',    'right_wrist'),
    'left_shoulder':  ('left_hip',       'left_shoulder',  'left_elbow'),
    'right_shoulder': ('right_hip',      'right_shoulder', 'right_elbow'),
    'left_ankle':     ('left_knee',      'left_ankle',     'left_foot_index'),
    'right_ankle':    ('right_knee',     'right_ankle',    'right_foot_index'),
}

MIN_VISIBILITY = 0.4  # ignore joints the model isn't confident about

# Spine needs a stricter bar than the 3-point angles above: it's computed
# from just two real landmarks (shoulder+hip) against a synthetic reference
# point, with no third real landmark to anchor it — more sensitive to a
# landmark's *position* being unreliable even while visibility still clears
# 0.4 (confirmed live 2026-08-27 in the TS port, frontend/src/pose/angles.ts).
MIN_VISIBILITY_SPINE = 0.6

# How far "up" (in normalized image-space y) the synthetic vertical
# reference point sits above the hip. The exact magnitude doesn't affect the
# resulting angle at all — _angle_deg only uses vector *direction*, not
# length — so this just needs to be a small positive number.
SPINE_UP_DELTA = 0.1


def _angle_deg(a: tuple, b: tuple, c: tuple) -> float:
    """Angle at b in the A-B-C triangle, using only x/y (2D)."""
    a2 = np.array(a[:2])
    b2 = np.array(b[:2])
    c2 = np.array(c[:2])
    ba = a2 - b2
    bc = c2 - b2
    norm = np.linalg.norm(ba) * np.linalg.norm(bc)
    if norm < 1e-9:
        return 0.0
    cos = np.clip(np.dot(ba, bc) / norm, -1.0, 1.0)
    return float(np.degrees(np.arccos(cos)))


def compute_angles(frame: PoseFrame) -> Dict[str, float]:
    """Return {angle_name: degrees} for all angles visible in this frame."""
    result: Dict[str, float] = {}
    lm  = frame.landmarks
    vis = frame.visibility

    for name, (a, b, c) in ANGLE_DEFS.items():
        if all(vis.get(j, 0.0) >= MIN_VISIBILITY for j in (a, b, c)):
            result[name] = _angle_deg(lm[a], lm[b], lm[c])

    # Spine: angle at the hip between the shoulder and a synthetic
    # straight-up reference point — normalized 2D, like every other angle
    # above (added 2026-08-30; previously computed only in the TS live path,
    # in pixel-space 3D against the original desktop pipeline's fixed-50px
    # reference — never measured from training data, just a hand-picked
    # threshold, which turned out too tight for real bodies on two exercises
    # live at the gym). Ported here so it can finally get a real measured
    # range the same way every other joint does, via write_exercise_angles.
    for side in ('right', 'left'):
        shoulder, hip = f'{side}_shoulder', f'{side}_hip'
        if vis.get(shoulder, 0.0) >= MIN_VISIBILITY_SPINE and vis.get(hip, 0.0) >= MIN_VISIBILITY_SPINE:
            hx, hy = lm[hip][0], lm[hip][1]
            vertical = (hx, hy - SPINE_UP_DELTA, 0.0)
            result['spine'] = _angle_deg(lm[shoulder], lm[hip], vertical)
            break

    return result


def frame_from_live(landmarks: Dict[int, object], width: int, height: int) -> PoseFrame:
    """
    Adapt the live detector's output (index → Point in pixel coords) to a
    PoseFrame, so the live pipeline computes angles with the exact same math
    and coordinate space (normalized 0-1) used to train the phase models.
    """
    lms: Dict[str, tuple] = {}
    vis: Dict[str, float] = {}
    for idx, name in LANDMARK_NAMES.items():
        pt = landmarks.get(idx)
        if pt is None:
            continue
        lms[name] = (pt.x / width, pt.y / height, getattr(pt, 'z', 0.0) / width)
        vis[name] = getattr(pt, 'visibility', 1.0)
    return PoseFrame(frame_idx=-1, timestamp_ms=0.0, landmarks=lms, visibility=vis)


def angles_over_time(frames: List[PoseFrame]) -> Dict[str, List[float]]:
    """
    Compute angles for every frame and pivot to {joint: [value_per_frame]}.
    Only joints visible in ALL frames are returned (avoids ragged arrays).
    """
    per_frame = [compute_angles(f) for f in frames]

    # Intersect keys so every joint has the same number of samples
    if not per_frame:
        return {}
    common = set(per_frame[0].keys())
    for d in per_frame[1:]:
        common &= d.keys()

    return {joint: [d[joint] for d in per_frame] for joint in sorted(common)}

File: frontend/src/pose/angles.ts
// Pose landmarks → joint angles (degrees).
// Faithful port of core/ml/angles.py — all 11 angles (10 joints + spine) are
// normalized-2D, the space the models and the Mongo ranges are measured in.

import type { Landmarks, LandmarkName } from './landmarks';

// (joint_a, vertex, joint_c) — angle measured at vertex. Mirrors ANGLE_DEFS.
const ANGLE_DEFS: Record<string, [LandmarkName, LandmarkName, LandmarkName]> = {
  left_knee:      ['left_hip',       'left_knee',      'left_ankle'],
  right_knee:     ['right_hip',      'right_knee',     'right_ankle'],
  left_hip:       ['left_shoulder',  'left_hip',       'left_knee'],
  right_hip:      ['right_shoulder', 'right_hip',      'right_knee'],
  left_elbow:     ['left_shoulder',  'left_elbow',     'left_wrist'],
  right_elbow:    ['right_shoulder', 'right_elbow',    'right_wrist'],
  left_shoulder:  ['left_hip',       'left_shoulder',  'left_elbow'],
  right_shoulder: ['right_hip',      'right_shoulder', 'right_elbow'],
  left_ankle:     ['left_knee',      'left_ankle',     'left_foot_index'],
  right_ankle:    ['right_knee',     'right_ankle',    'right_foot_index'],
};

const MIN_VISIBILITY = 0.4;

// Spine needs a stricter bar than the 3-point joint angles above: it's
// computed from just two landmarks (shoulder+hip) against a synthetic
// "straight up" reference point, with no third real landmark to anchor it —
// far more sensitive to a landmark's *position* being unreliable even while
// its visibility score still clears 0.4. Confirmed live 2026-08-27: as a
// user nears the frame edge, one side's shoulder/hip stayed just above 0.4
// while its estimated position was already distorted by perspective, so
// `spine` spiked and false-triggered the (global, high-severity, spoken)
// "don't arch your back" correction on every frame exit, not on real
// posture. Applies only to spine's own two landmarks, not the shared
// MIN_VISIBILITY gate other angles use.
const MIN_VISIBILITY_SPINE = 0.6;

// How far "up" (normalized image-space y) the synthetic vertical reference
// point sits above the hip. Doesn't affect the resulting angle at all —
// angle2d only uses vector *direction*, not length — just needs to be a
// small positive number. Mirrors core/ml/angles.py exactly.
const SPINE_UP_DELTA = 0.1;

// Angle at b in the A-B-C triangle, using x/y only (normalized 0-1 screen
// space) — the space all 10 joint angles below are measured in.
function angle2d(
  a: { x: number; y: number }, b: { x: number; y: number }, c: { x: number; y: number },
): number {
  const bax = a.x - b.x, bay = a.y - b.y;
  const bcx = c.x - b.x, bcy = c.y - b.y;
  const norm = Math.hypot(bax, bay) * Math.hypot(bcx, bcy);
  if (norm < 1e-9) return 0;
  const cos = Math.min(1, Math.max(-1, (bax * bcx + bay * bcy) / norm));
  return (Math.acos(cos) * 180) / Math.PI;
}

/** All joint angles visible in this frame, keyed by angle name. */
export function computeAngles(lms: Landmarks): Record<string, number> {
  const out: Record<string, number> = {};

  for (const [name, [a, b, c]] of Object.entries(ANGLE_DEFS)) {
    const pa = lms[a], pb = lms[b], pc = lms[c];
    if (!pa || !pb || !pc) continue;   // joint off-frame this tick — just skip it
    if (pa.visibility < MIN_VISIBILITY || pb.visibility < MIN_VISIBILITY ||
        pc.visibility < MIN_VISIBILITY) continue;   // MediaPipe isn't confident enough
    out[name] = angle2d(pa, pb, pc);
  }

  // Spine: angle at the hip between the shoulder and a synthetic straight-up
  // reference point — normalized 2D like everything else above (changed
  // 2026-08-30 from pixel-space 3D against a fixed-50px reference; see
  // core/ml/angles.py's mirror of this for why: that version could never be
  // measured from training data at all, since the training pipeline only
  // ever sees normalized landmark coordinates, no pixel dimensions).
  for (const side of ['right', 'left'] as const) {
    const shoulder = lms[`${side}_shoulder`];
    const hip = lms[`${side}_hip`];
    if (!shoulder || !hip) continue;
    if (shoulder.visibility < MIN_VISIBILITY_SPINE || hip.visibility < MIN_VISIBILITY_SPINE) continue;
    const vertical = { x: hip.x, y: hip.y - SPINE_UP_DELTA };
    out.spine = angle2d(shoulder, hip, vertical);
    break;
  }

  return out;
}
